zenoh/src/api/session.rs

```diff
@@ -359,142 +367,120 @@ impl Resource {
-            #[cfg(feature = "unstable")]
-            reliability: Reliability::DEFAULT,
-            destination: Locality::default(),
-        }
+impl<'a, T> UndeclarableSealed<&'a Session> for T
+where
+    T: UndeclarableSealed<()>,
+{
+    type Undeclaration = <T as UndeclarableSealed<()>>::Undeclaration;
+
+    fn undeclare_inner(self, _session: &'a Session) -> Self::Undeclaration {
+        self.undeclare_inner(())
     }
-    #[zenoh_macros::unstable]
-    fn liveliness(&'s self) -> Liveliness<'a> {
-        Liveliness {
-            session: self.clone(),
-        }
+}
+
+// NOTE: `UndeclarableInner` is only pub(crate) to hide the `undeclare_inner` method. So we don't
+// care about the `private_bounds` lint in this particular case.
+#[allow(private_bounds)]
+/// A trait implemented by types that can be undeclared.
+pub trait Undeclarable<S = ()>: UndeclarableSealed<S> {}
+
+impl<T, S> Undeclarable<S> for T where T: UndeclarableSealed<S> {}
+
+pub(crate) struct SessionInner {
+    weak_counter: Mutex<usize>,
+    pub(crate) runtime: Runtime,
+    pub(crate) state: RwLock<SessionState>,
+    pub(crate) id: u16,
+    owns_runtime: bool,
+    task_controller: TaskController,
+}
+
+impl fmt::Debug for SessionInner {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_struct("Session")
+            .field("id", &self.runtime.zid())
+            .finish()
     }
-    fn info(&'s self) -> SessionInfo<'a> {
-        SessionInfo {
-            session: self.clone(),
-        }
+}
+
+/// A zenoh session.
+///
+pub struct Session(pub(crate) Arc<SessionInner>);
+
+impl Session {
+    pub(crate) fn downgrade(&self) -> WeakSession {
+        WeakSession::new(&self.0)
     }
 }
 
-impl Deref for SessionRef<'_> {
-    type Target = Session;
+impl fmt::Debug for Session {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        self.0.fmt(f)
+    }
+}
 
-    fn deref(&self) -> &Self::Target {
-        match self {
-            SessionRef::Borrow(b) => b,
-            SessionRef::Shared(s) => s,
-        }
+impl Clone for Session {
+    fn clone(&self) -> Self {
+        let _weak = self.0.weak_counter.lock().unwrap();
+        Self(self.0.clone())
     }
 }
 
-impl fmt::Debug for SessionRef<'_> {
-    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        match self {
-            SessionRef::Borrow(b) => Session::fmt(b, f),
-            SessionRef::Shared(s) => Session::fmt(s, f),
+impl Drop for Session {
+    fn drop(&mut self) {
+        let weak = self.0.weak_counter.lock().unwrap();
+        if Arc::strong_count(&self.0) == *weak + 1 {
+            drop(weak);
+            if let Err(error) = self.close().wait() {
+                tracing::error!(error)
+            }
         }
     }
 }
 
-pub(crate) trait UndeclarableSealed<S, O, T = ZResult<()>>
-where
-    O: Resolve<T> + Send,
-{
-    fn undeclare_inner(self, session: S) -> O;
+pub(crate) struct WeakSession(Arc<SessionInner>);
```

Comment:
Why `WeakSession` contains an `Arc` instead of `Weak`? I find it a beat confusing.

Reply:
At first, I used Weak everywhere (as you can see in the commit history), because it was simpler for a first iteration. However, I also add to use Weak for the primitives, and it there was the rub...
When the session was dropped, primitive was supposed to be still usable for the closing, and using Weak here was causing chain reaction leading to panic in runtime.
I didn't wanted to investigate to much, because it seemed complex, and I already had the idea in my mind to replace Weak by a pseudo-weak reference but using Arc. I mean, Weak means worse performance, notably for Publisher and impl Primitive, so if we have a simple way to avoid it, we should do it. But here, I also had to do it to avoid the panic.
So I implemented this "pseudo-weak", using an additional counter in a mutex. Why? because using atomic seemed very complicated to me — what happens when pseudo-weak are cloned when the last Session is dropped?
So I went with this simple but powerful implementation, bringing the pseudo-weak semantic we need for session closing, and the guaranteeing the best performance for the critical paths.
The mutex only gives a tiny overload for clone/drop operations, but we are talking about session entities, so I assume we don't care at all here.

Reply:
Still, your question highlight the lack of comments in the code to explain the approach. I will have to write them.

Reply:
It sounds reasonable, thanks!